pydantic/_internal/_generate_schema.py

```diff
@@ -766,3 +771,137 @@ def get_model_self_schema(cls: type[BaseModel]) -> core_schema.ModelSchema:
+    )
+
+    if nullable_schema is None:
+        return discriminated_schema
+    else:
+        return core_schema.nullable_schema(
+            discriminated_schema,
+            strict=nullable_schema.get('strict'),
+            ref=nullable_schema.get('ref'),
+            metadata=nullable_schema.get('metadata'),
+            serialization=nullable_schema.get('serialization'),
+        )
+
+
+def _build_tagged_union_choices(
+    choices: typing.Iterable[core_schema.CoreSchema], discriminator: str, aliases: dict[str, None]
+) -> dict[str | int, str | int | core_schema.CoreSchema]:
+    choices = list(choices)[::-1]
+
+    tagged_union_choices: dict[str | int, str | int | core_schema.CoreSchema] = {}
+    while choices:
+        choice = choices.pop()
+        if choice['type'] == 'union':
+            choices.extend(choice['choices'])
+            continue
+
+        discriminator_values = _get_discriminator_values_for_choice(choice, discriminator, aliases)
+        if discriminator_values:
+
+            def _handle_discriminator_value(value: Any, choice_override: int | str | None = None) -> None:
+                # This function accepts choice_override so that we can produce a schema that doesn't copy choices
+                if not isinstance(value, (int, str, Enum)):
+                    raise ValueError(f'Invalid discriminator value {value!r}; must be a string, int, or Enum')
+                if isinstance(value, Enum):
+                    value = value.value
+                if value in tagged_union_choices:
+                    # Need to walk the choices dict until we get to a "real" choice
+                    existing_choice = tagged_union_choices[value]
+                    while isinstance(existing_choice, (str, int)):
+                        existing_choice = tagged_union_choices[existing_choice]
+                    if existing_choice != choice:
+                        raise ValueError(
+                            f'Value {value!r} for discriminator {discriminator!r} mapped to multiple choices'
+                        )
+                else:
+                    tagged_union_choices[value] = choice if choice_override is None else choice_override
+
+            primary_value = discriminator_values[0]
+            _handle_discriminator_value(primary_value)
+            for other_value in discriminator_values[1:]:
+                _handle_discriminator_value(other_value, primary_value)
+    return tagged_union_choices
+
+
+def _get_discriminator_values_for_choice(
+    choice: core_schema.CoreSchema, discriminator: str, aliases: dict[str, None]
+) -> list[Any]:
+    if choice['type'] == 'tagged-union':
+        values: list[Any] = []
+        for inner_choice in choice['choices'].values():
+            if isinstance(inner_choice, (str, int)):
+                continue
+            values.extend(_get_discriminator_values_for_choice(inner_choice, discriminator, aliases))
+        return values
```

Comment:
Do we also need to handle the case of a nullable schema here? Any other cases we can "collapse" into a union? I'm not sure if it happens earlier but I wonder how we handle:

```python
Result =T | Exception
MaybeIntResult = Result[int] | None
```

Reply:
good catch. I'll see if I can produce an error in a new test with the current code.